internal/throttler/remote/throttler.go

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2018 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"fmt"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/uber-go/atomic"
+
+	"github.com/uber/jaeger-client-go"
+	"github.com/uber/jaeger-client-go/utils"
+)
+
+const (
+	// minimumCredits is the minimum amount of credits necessary to not be throttled.
+	// i.e. if currentCredits > minimumCredits, then the operation will not be throttled.
+	minimumCredits = 1.0
+)
+
+type creditResponse struct {
+	Operation string  `json:"operation"`
+	Credits   float64 `json:"credits"`
+}
+
+type httpCreditManagerProxy struct {
+	hostPort string
+	logger   jaeger.Logger
+}
+
+func newHTTPCreditManagerProxy(hostPort string, logger jaeger.Logger) *httpCreditManagerProxy {
+	return &httpCreditManagerProxy{
+		hostPort: hostPort,
+		logger:   logger,
+	}
+}
+
+func (m *httpCreditManagerProxy) FetchCredits(uuid, serviceName string, operation []string) []creditResponse {
+	params := url.Values{}
+	params.Set("service", serviceName)
+	params.Set("uuid", uuid)
+	for _, op := range operation {
+		params.Add("operation", op)
+	}
+	var resp []creditResponse
+	if err := utils.GetJSON(fmt.Sprintf("http://%s/credits?%s", m.hostPort, params.Encode()), &resp); err != nil {
+		m.logger.Error("Failed to receive credits from agent: " + err.Error())
+	}
+	return resp
+}
+
+// Throttler retrieves credits from agent and uses it to throttle operations.
+type Throttler struct {
+	options
+
+	mux           sync.RWMutex
```

Comment:
For the record, I used to love the idea of reader-writer locks in all code, then learned that it might actually perform worse than a regular mutex if the contention isn't high enough. RW locks generally have a bit more overhead because the logic to avoid writer/reader starvation is more complex. Might be worth considering and/or profiling.